cmd/libsnap-confine-private/snap-dir.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include "snap-dir.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "snap.h"  // For SC_SNAP_MOUNT_DIR_UNSUPPORTED
+
+static const char *_snap_mount_dir = NULL;
+
+const char *sc_snap_mount_dir(sc_error **errorp) {
+    sc_error *err = NULL;
+
+    if (_snap_mount_dir == NULL) {
+        err = sc_error_init_api_misuse("sc_probe_snap_mount_dir was not called yet");
+    }
+
+    sc_error_forward(errorp, err);
+    return _snap_mount_dir;
+}
+
+#define SC_PROC_1_ROOT "/proc/1/root"
+
+void sc_probe_snap_mount_dir(int root_fd, sc_error **errorp) {
+    char target[PATH_MAX];
+    struct stat sb;
+    ssize_t n;
+    sc_error *err = NULL;
+
+    const char *const probe_dir = SC_PROC_1_ROOT SC_PREFERRED_SNAP_MOUNT_DIR + (root_fd == AT_FDCWD ? 0 : 1);
```

Comment:
this is not easy to follow, how about
```suggestion
#define SC_PROC_1_ROOT_REL "proc/1/root"    

    // absolute path to the preferred snap mount dir, unless probing relative 
    // to the current work dir
    const char *probe_dir = "/" SC_PROC_1_ROOT_REL SC_PREFERRED_SNAP_MOUNT_DIR;
    if (root_fd == AT_FDCWD) {
        probe_dir = SC_PROC_1_ROOT_REL SC_PREFERRED_SNAP_MOUNT_DIR;
    }
```